synapse/media/oembed.py

```diff
@@ -109,18 +109,20 @@ def autodiscover_from_html(self, tree: "etree.Element") -> Optional[str]:
             The URL to use for oEmbed information, or None if no URL was found.
         """
         # Search for link elements with the proper rel and type attributes.
-        for tag in tree.xpath(
-            "//link[@rel='alternate'][@type='application/json+oembed']"
+        for tag in cast(
+            List["etree._Element"],
+            tree.xpath("//link[@rel='alternate'][@type='application/json+oembed']"),
         ):
```

Comment:
See https://lxml.de/xpathxslt.html#xpath-return-values

tl;dr `xpath` returns a massive `Union` of types, what type is returned is determined by the expression itself. We could use asserts instead of casts though if that seems better.

Reply:
For every single instance?

Reply:
Oh, sorry---there are quite a few of them. In that case it's probably fine to leave it uncommented (I guess our future selves can always rediscover this PR.) 

(Though it seems I responded too slowly. I'll let you keep or remove the comments at your discretion.)